pkg/psalabelsyncer/podsecurity_label_sync_controller.go

```diff
@@ -0,0 +1,285 @@
+	securityv1listers "github.com/openshift/client-go/security/listers/security/v1"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+const (
+	controllerName        = "pod-security-admission-label-synchronization-controller"
+	labelSyncControlLabel = "security.openshift.io/scc.podSecurityLabelSync"
+)
+
+// PodSecurityAdmissionLabelSynchronizationController watches over namespaces labelled with
+// "security.openshift.io/scc.podSecurityLabelSync: true" and configures the PodSecurity
+// admission namespace label to match the user account privileges in terms of being able
+// to use SCCs
+type PodSecurityAdmissionLabelSynchronizationController struct {
+	namespaceClient corev1client.NamespaceInterface
+
+	namespaceLister          corev1listers.NamespaceLister
+	roleLister               rbacv1listers.RoleLister
+	roleBindingLister        rbacv1listers.RoleBindingLister
+	clusterRoleLister        rbacv1listers.ClusterRoleLister
+	clusterRoleBindingLister rbacv1listers.ClusterRoleBindingLister
+	serviceAccountLister     corev1listers.ServiceAccountLister
+	sccLister                securityv1listers.SecurityContextConstraintsLister
+
+	nsLabelSelector labels.Selector
+
+	saToSCCsCache *SAToSCCCache
+}
+
+func NewPodSecurityAdmissionLabelSynchronizationController(
+	namespaceClient corev1client.NamespaceInterface,
+	namespaceInformer corev1informers.NamespaceInformer,
+	rbacInformers rbacv1informers.Interface,
+	serviceAccountInformer corev1informers.ServiceAccountInformer,
+	sccInformer securityv1informers.SecurityContextConstraintsInformer,
+	eventRecorder events.Recorder,
+) (factory.Controller, error) {
+
+	// add the indexers that are used in the SAToSCC cache
+	if err := rbacInformers.RoleBindings().Informer().AddIndexers(
+		cache.Indexers{BySAIndexName: BySAIndexKeys},
+	); err != nil {
+		return nil, err
+	}
+
+	if err := rbacInformers.ClusterRoleBindings().Informer().AddIndexers(
+		cache.Indexers{BySAIndexName: BySAIndexKeys},
+	); err != nil {
+		return nil, err
+	}
+
+	controlledNamespacesLabelSelector, err := controlledNamespacesLabelSelector()
+	if err != nil {
+		return nil, err
+	}
+
+	c := &PodSecurityAdmissionLabelSynchronizationController{
+		namespaceClient: namespaceClient,
+
+		namespaceLister:          namespaceInformer.Lister(),
+		roleLister:               rbacInformers.Roles().Lister(),
+		roleBindingLister:        rbacInformers.RoleBindings().Lister(),
+		clusterRoleLister:        rbacInformers.ClusterRoles().Lister(),
+		clusterRoleBindingLister: rbacInformers.ClusterRoleBindings().Lister(),
+		serviceAccountLister:     serviceAccountInformer.Lister(),
+		sccLister:                sccInformer.Lister(),
+
+		nsLabelSelector: controlledNamespacesLabelSelector,
+
+		saToSCCsCache: NewSAToSCCCache(rbacInformers, sccInformer),
+	}
+
+	return factory.New().
+		WithSync(c.sync).
+		WithFilteredEventsInformersQueueKeysFunc(
```

Comment:
do you guys prefer the way these factor out while using this factory thing or do you prefer "normal" construction?

Reply:
I think the factoring is making it easier to observe what the expectations of the controller might be. Although I haven't written many "normal" kube-controllers, perhaps there would be a way to keep the inputs well readable, too.